proposals/build-provenance-and-code-signing-for-homebrew.md

```diff
@@ -0,0 +1,112 @@
+
+Because Homebrew uses GitHub Actions for all `homebrew-core` builds, the Homebrew CI has trivial access to GitHub Actions' [ambient OpenID Connect identities](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect). These identities can in turn be used to generate build provenance and sign with Sigstore, meaning that Homebrew's existing CI workflows would require relatively little additional configuration (and zero external key management) to produce valid SLSA attestations and Sigstore signatures for each bottle build.
+
+The high-level flow for bottle signing:
+
+1. Existing bottle-building workflows will gain a signing step, which will only run as a finalization step for bottles that appear on `master` (i.e., not just the temporary bottles that are built as part of the normal PR lifecycle).
+2. The signing step is responsible for producing an attestation with a Sigstore signature that _attests_ to the bottle. This signed attestation should include, at minimum, the bottle archive's SHA256 digest and the bottle's filename. The digest is necessary to produce a proof that the signature is for a particular bottle's contents, and the filename is necessary to produce a proof that the signature is intended to accompany a particular install step (i.e. `brew install foo` should not verify if an attacker substitutes `foo`'s bottle for a correctly signed bottle of `bar`, or even a different version of `foo` than the resolver produces). The attestation will follow the [in-toto format](https://github.com/in-toto/attestation/tree/main/spec/v1) and thus will be consistent with [SLSA's attestation model](https://slsa.dev/attestation-model).
+3. The signing step is responsible for uploading the resulting signature to the same object storage as the bottle, and ensuring that the uploaded signature is addressable in a predictable and publicly derivable manner from just the formula and bottle's own metadata, [consistent with SLSA's guidenlines on distributing provenance](https://slsa.dev/spec/v1.0/distributing-provenance). In other words: a client that knows how to access the bottle's resource should also be able to access the bottle's signature resource with no additional metadata, similar to how [signature location and management](https://docs.sigstore.dev/cosign/signing_with_containers/#signature-location-and-management) works for signing container images with Sigstore.
+
+At the end of this flow, both the bottle and its attestation with signature should appear in the tap's associated object storage.
```

Comment:
Could you clarify where this object storage would be exactly?

Reply:
That partially depends on @steiza 🙂 

The answer here is not particularly satisfying: third-party taps can put their object storage wherever they please, and Homebrew isn't prescriptive about that (other than having support for a few download strategies already built in).

The "simpler" answer is that this proposal is initially targeted at `homebrew-core`, and `homebrew-core` *currently* repurposes GitHub Packages to do object storage for its bottles. However, that's an implementation detail, so I'd prefer to leave that out (in the interest of keeping the proposal mostly behavioral in structure). 

That being said: if you think it'll be helpful to add a note about *specific* implementation possibilities, I can do that!

Reply:
Agreed: the exact choice of object storage doesn't matter, so long as the threat model covers that, no matter what, it can be compromised and tampered with.

Reply:
Cool; let me know if you think the language in #25 covers that!

Reply:
Yes, I think it's covered by "Model 1: Compromise of bottle storage". Thanks!

Reply:
Resolved per #25.